evm/p2p/kademlia.py

```diff
@@ -0,0 +1,581 @@
+        fails to respond to a ping.
+        """
+        self.last_updated = time.time()
+        if node in self.nodes:
+            self.nodes.remove(node)
+            self.nodes.append(node)
+        elif len(self) < self.k:
+            self.nodes.append(node)
+        else:
+            self.replacement_cache.append(node)
+            return self.head
+
+    @property
+    def head(self):
+        """Least recently seen"""
+        return self.nodes[0]
+
+    @property
+    def depth(self):
+        """Depth is the prefix shared by all nodes in the bucket.
+
+        i.e. The number of shared leading bits.
+        """
+        def to_binary(x):  # left padded bit representation
+            b = bin(x)[2:]
+            return '0' * (k_id_size - len(b)) + b
+
+        if len(self.nodes) < 2:
+            return k_id_size
+
+        bits = [to_binary(n.id) for n in self.nodes]
+        for i in range(1, k_id_size + 1):
+            if len(set(b[:i] for b in bits)) != 1:
+                return i - 1
+        # This means we have at least two nodes with the same ID, so raise an AssertionError
+        # because we don't want it to be caught accidentally.
+        raise AssertionError("Unable to calculate depth")
+
+    def __contains__(self, node):
+        return node in self.nodes
+
+    def __len__(self):
+        return len(self.nodes)
+
+
+class RoutingTable():
+
+    def __init__(self, node):
+        self.this_node = node
+        self.buckets = [KBucket(0, k_max_node_id)]
+
+    def split_bucket(self, bucket):
+        a, b = bucket.split()
+        index = self.buckets.index(bucket)
+        self.buckets[index] = a
+        self.buckets.insert(index + 1, b)
+
+    @property
+    def idle_buckets(self):
+        idle_cutoff_time = time.time() - k_idle_bucket_refresh_interval
+        return [b for b in self.buckets if b.last_updated < idle_cutoff_time]
+
+    @property
+    def not_full_buckets(self):
+        return [b for b in self.buckets if not b.is_full]
+
+    def remove_node(self, node):
+        self.bucket_by_node(node).remove_node(node)
+
+    def add_node(self, node):
+        if node == self.this_node:
+            raise ValueError("Cannot add this_node to routing table")
+        bucket = self.bucket_by_node(node)
+        eviction_candidate = bucket.add(node)
+        if eviction_candidate is not None:  # bucket is full
+            # Split if the bucket has the local node in its range or if the depth is not congruent
+            # to 0 mod k_b
+            depth = bucket.depth
+            if bucket.in_range(self.this_node) or (depth % k_b != 0 and depth != k_id_size):
+                self.split_bucket(bucket)
+                return self.add_node(node)  # retry
+            # Nothing added, ping eviction_candidate
+            return eviction_candidate
+        return None  # successfully added to not full bucket
+
+    def bucket_by_node(self, node):
+        for bucket in self.buckets:
```

Comment:
The number of buckets will always be relatively small, so I don't think the extra complexity of a binary search is warranted here.

Reply:
Out of curiosity, when you say small does that mean 10/100/500?  Also, I assume you know it exists, but does that hold even with the existence of the `bisect` library which provides built-in implementations?

Reply:
I believe they'd stay on the order of hundreds. Geth uses a different implementation with a fixed list of 256 buckets, for instance.
As for using bisect here, we'd need to implement the sort-related methods in Bucket, no? That's why I think it may not be warranted, but I'm happy to give it a try if you feel strong about it

Reply:
I think I do feel strongly enough to advocate for this being implemented **but** I think it's a good candidate for an *easy pickins* ticket so I'd suggest not doing it and I'll create a ticket.